DTC warning suppression

agansari opened the pull request:
Allows boards to have a extra.cmake that allows for custom cmake commands.
We use the extra.cmake file to set EXTRA_DTC_FLAGS to suppress warnings on LPC SoC that have a single register address for all gpio ports.

Fixes #20748

agansari committed:
cmake: allow boards to have extra cmake parametes

This commit allows boards to have custom cmake parameters that are
loaded before device tree is processed.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
boards: suppress dtc simple_bus_reg warning on LPC

LPC SoCs share the same register for all GPIO poarts. This leads us to a
device tree configuration that missmatches GPIO register address and
GPIO port identifier.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
scripts: simple-bus warning suppression

Single-bus warning in python parsing of device tree is suppressed if
this is also suppresed in the device tree compiler.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -724,9 +724,19 @@ def unit_addr(self):
+            EXTRA_DTC_FLAGS = ""
+
+        # This code is redundant it checks the same thing as simple_bus_reg in dtc
+        # disable it if it's suppressed in dtc
```
Could end sentences in a period to make this a bit easier to read.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -724,9 +724,19 @@ def unit_addr(self):
+            if EXTRA_DTC_FLAGS is None:
+                EXTRA_DTC_FLAGS = ""
+        except:
+            EXTRA_DTC_FLAGS = ""
```
Can shorten this down to this:

```python
extra_dtc_flags = os.getenv("EXTRA_DTC_FLAGS", "")
```

`""` is the fallback value.

The `except` would never hit, because `get()` never raises.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -724,9 +724,19 @@ def unit_addr(self):
+
+        # This code is redundant it checks the same thing as simple_bus_reg in dtc
+        # disable it if it's suppressed in dtc
+        if EXTRA_DTC_FLAGS.find('no-simple_bus_reg') == -1:
```
Can do

```python
if "-Wno-simple_bus_reg" in extra_dtc_flags:
```

Could combine it all into something like

```python
if "-Wno-simple_bus_reg" not in os.getenv("EXTRA_DTC_FLAGS", ""):
```

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -724,9 +724,19 @@ def unit_addr(self):
+        if EXTRA_DTC_FLAGS.find('no-simple_bus_reg') == -1:
+            if self.regs and self.regs[0].addr != addr :
+                self.edt._warn("unit-address and first reg (0x{:x}) don't match "
+                            "for {}".format(self.regs[0].addr, self.name))
```
Broken alignment.

ulfalizer replied:
Also extra space in `... != addr :`.

ulfalizer commented on boards/arm/lpcxpresso54114/extra.cmake:
```diff
@@ -0,0 +1,11 @@
+
+
+# Suppress "simple_bus_reg" on LPC boards as all GPIO ports use the same register.
+set(EXTRA_DTC_FLAGS "-Wno-simple_bus_reg")
```
11 lines for 1 line of code might be a bit long. Can make it

```cmake
# Copyright (c) 2019, NXP
# SPDX-License-Identifier: Apache-2.0

# Suppress "simple_bus_reg" on LPC boards as all GPIO ports use the same register
set(EXTRA_DTC_FLAGS "-Wno-simple_bus_reg")
```

if the copyright/license stuff needs to be there.

agansari replied:
agree

ulfalizer commented on cmake/app/boilerplate.cmake:
```diff
@@ -442,6 +442,11 @@ set(CMAKE_CXX_COMPILER_FORCED 1)
+# Include board specific dtc flags and export them to python via system
+# variables to where device tree is parsed.
+include(${BOARD_DIR}/extra.cmake OPTIONAL)
+set(ENV{EXTRA_DTC_FLAGS} "${EXTRA_DTC_FLAGS}")
```
I wonder if this mechanism might be a bit general for just changing a single flag. Feels like it might get abused.

@SebastianBoe 
Is there some other way to deal with this, if needed?

SebastianBoe replied:
It is overly general, but there is no better way. In practice, this has not proven to be a problem for board.cmake, and other hooks.

agansari replied:
@ulfalizer @SebastianBoe i didn't find a better way to pass the information to python. Will leave it as is since it's similar to board.cmake

ulfalizer commented on boards/arm/lpcxpresso55s69/extra.cmake:
```diff
@@ -0,0 +1,11 @@
+
+
+# Suppress "simple_bus_reg" on LPC boards as all GPIO ports use the same register.
+set(EXTRA_DTC_FLAGS "-Wno-simple_bus_reg")
```
Ditto here.

ulfalizer reviewed (commented):
Wonder if there's some simpler way to fix it on the devicetree side, e.g. by leaving out the unit address, or refactoring in some other way.

SebastianBoe commented:
What's wrong with the board.cmake hook? Is it placed too early / too late ?

agansari commented:
> What's wrong with the board.cmake hook? Is it placed too early / too late ?

It's too late, my first goto was board.cmake but device tree is ready compiled before EXTRA_DTC_FLAGS is set.

SebastianBoe commented on scripts/dts/edtlib.py:
```diff
@@ -724,9 +724,19 @@ def unit_addr(self):
-            self.edt._warn("unit-address and first reg (0x{:x}) don't match "
-                           "for {}".format(self.regs[0].addr, self.name))
+        try:
+            EXTRA_DTC_FLAGS = os.environ.get("EXTRA_DTC_FLAGS")
```
We should not use env vars to communicate information between the build system and underlying scripts. CLI arguments should be used.

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -442,6 +442,11 @@ set(CMAKE_CXX_COMPILER_FORCED 1)
 
+# Include board specific dtc flags and export them to python via system
+# variables to where device tree is parsed.
+include(${BOARD_DIR}/extra.cmake OPTIONAL)
```
let's call it pre_dt_board.cmake to be consistent with board.cmake.

SebastianBoe replied:
needs documentation.

SebastianBoe commented on boards/arm/lpcxpresso54114/extra.cmake:
```diff
@@ -0,0 +1,11 @@
+
+
+# Suppress "simple_bus_reg" on LPC boards as all GPIO ports use the same register.
+set(EXTRA_DTC_FLAGS "-Wno-simple_bus_reg")
```
list(APPEND

so we don't overwrite any previously set value.

agansari replied:
@SebastianBoe i see no point to appending, EXTRA_DTC_FLAGS is not used anywhere else, it's there to allow flexibility, but was never set.

SebastianBoe replied:
It could be set by apps for instance.

agansari replied:
okay will append

agansari commented:
@SebastianBoe @ulfalizer made the changes required, i still need to document pre_dt_board.cmake

SebastianBoe commented:
@agansari : This feedback has not been addressed:

> We should not use env vars to communicate information between the build system and underlying scripts. CLI arguments should be used.

agansari committed:
cmake: pass extra dtc flags to python scripts

Pass arguments as command line from cmake to python. Extra DTC flags are
used later on to check errors in python.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
doc: device tree - document pre_dt_board.cmake

Document pre_dt_board.cmake file usage.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari commented:
@SebastianBoe @ulfalizer updated documentation and CLI based parameter passing

SebastianBoe commented on doc/guides/dts/index.rst:
```diff
@@ -712,6 +712,14 @@ generated in the same directory concatenating all board-related fixup files.
 files, giving Zephyr C source files access to the board's devicetree information.
 
+Since some boards may need special instructions to the device tree compiler
+(e.g. warning suppression) you can add extra flags in ``pre_dt_board.cmake``.
```
We need to say where pre_dt_board.cmake is located.

agansari replied:
Updated

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -712,6 +712,15 @@ generated in the same directory concatenating all board-related fixup files.
 The include file ``include/generated_dts_board.h`` includes both these generated
 files, giving Zephyr C source files access to the board's devicetree information.
 
+Since some boards may need special instructions to the device tree compiler
```
change ``instructions to`` to ``instructions for``

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -712,6 +712,15 @@ generated in the same directory concatenating all board-related fixup files.
 
+Since some boards may need special instructions to the device tree compiler
+(e.g. warning suppression), create a file named ``pre_dt_board.cmake`` inside
+board's folder and add the extra flags inside.
```
change ``board's`` to ``the board's``

agansari replied:
@dbkinder fixed both

dbkinder reviewed (changes requested):
couple of grammar typos

dbkinder reviewed (approved):
+1 for doc changes, thanks!

galak commented on boards/arm/lpcxpresso54114/pre_dt_board.cmake:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Suppress "simple_bus_reg" on LPC boards as all GPIO ports use the same register.
+list(APPEND EXTRA_DTC_FLAGS "-Wno-simple_bus_reg")
```
Can we do this in the SoC cmake instead of the board one?

SebastianBoe replied:
We can not unfortunately. The SoC CMake is processed after DT.

agansari replied:
@galak I've tried it earlier but as @SebastianBoe mentioned SoC is processed after DT.
Also I think having it later on helps us have a finer granularity.
Fixing pylint next.

galak reviewed (changes requested):
Can you look and see if we can move setting EXTRA_DTC_FLAGS in the SoC Cmake instead.

Also can you fixup the pylint error.

agansari commented:
@galak fixed pylint, please review again
